### DISCUSSION TITLE
Implement a filter on update that matches on a subquery

### DISCUSSION THREAD

**Discussion by LtdJorge** [diesel-rs/diesel#3952] [Q&A] [answered]
I have the query on the left, and I'm banging my head on how to translate it to diesel. I am trying to update just one row, so I'm doing a subquery to find that row and pass that to filter. If debug both queries separately (filtering the outer query by id=1 instead of the subquery), I get the correct SQL for both. Here are the query, my diesel implementation (it's a bit messy because I'm using diesel_async) and the error that it gives: 
![image](https://github.com/diesel-rs/diesel/assets/23530149/e32c899b-f03c-443d-88b5-0b8dd1d9fc3a)

The table definition is included as a doc comment on the function.

The debugs of both queries separately are:
- Subquery: `(SELECT "codigos"."id" FROM "codigos" WHERE (("codigos"."expired" = $1) AND ("codigos"."expires_at" < date(CURRENT_TIMESTAMP))) LIMIT $2 FOR NO KEY UPDATE SKIP LOCKED) -- binds: [false, 1]`
- Outer query: ` UPDATE "codigos" SET "expired" = $1 WHERE ("codigos"."id" = $2) RETURNING "codigos"."id" -- binds: [true, 1]`

What am I doing wrong? If I need to provide more context, let me know.

**Comment by weiznich** [comment]
Can you provide a minimal reproducible example or at least the relevant source code as text. That would make it much easier for us to reproduce the problem on our own.

**Comment by LtdJorge** [comment]
Sure, uploaded to https://github.com/LtdJorge/diesel-reproducible-example, I confirmed the same exact error occurs.

**Reply by LtdJorge** [reply]
I left the postgres URL params as String::new() so that anyone can plug their own to test locally.

**Reply by LtdJorge** [reply]
I added the migrations and the query from the screenshot too

**Reply by LtdJorge** [reply]
I have found that it can be reduced to the following:

```rust
let unique_id = codigos
    .select(id)
    .single_value();

let updated = update(codigos.filter(id.nullable().eq(unique_id)));
```

Which gives the following error:
![image](https://github.com/diesel-rs/diesel/assets/23530149/568964ce-f1f3-425f-95c8-0fbe1cb5596c)
Text version:
```
error[E0277]: the trait bound `SelectStatement<FromClause<codigos::table>, query_builder::select_clause::DefaultSelectClause<FromClause<codigos::table>>, query_builder::distinct_clause::NoDistinctClause, query_builder::where_clause::WhereClause<diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<NullableExpression<codigos::columns::id>, diesel::expression::grouped::Grouped<diesel::expression::subselect::Subselect<SelectStatement<FromClause<codigos::table>, query_builder::select_clause::SelectClause<codigos::columns::id>, query_builder::distinct_clause::NoDistinctClause, query_builder::where_clause::NoWhereClause, query_builder::order_clause::NoOrderClause, LimitOffsetClause<LimitClause<diesel::expression::bound::Bound<BigInt, i64>>, NoOffsetClause>>, diesel::sql_types::Nullable<Integer>>>>>>>: IntoUpdateTarget` is not satisfied
  --> src/jobs/expire_timer.rs:37:34
   |
37 |             let updated = update(codigos.filter(id.nullable().eq(unique_id)));
   |                           ------ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `IntoUpdateTarget` is not implemented for `SelectStatement<FromClause<table>, DefaultSelectClause<FromClause<table>>, NoDistinctClause, WhereClause<...>>`
   |                           |
   |                           required by a bound introduced by this call
   |
   = help: the trait `IntoUpdateTarget` is implemented for `SelectStatement<FromClause<F>, query_builder::select_clause::DefaultSelectClause<FromClause<F>>, query_builder::distinct_clause::NoDistinctClause, W>`
note: required by a bound in `update`
  --> /home/jorge/.cargo/registry/src/index.crates.io-6f17d22bba15001f/diesel-2.1.4/src/query_builder/functions.rs:80:18
   |
80 | pub fn update<T: IntoUpdateTarget>(source: T) -> UpdateStatement<T::Table, T::WhereClause> {
   |                  ^^^^^^^^^^^^^^^^ required by this bound in `update`
help: consider removing this method call, as the receiver has type `codigos::table` and `codigos::table: IntoUpdateTarget` trivially holds
   |
37 -             let updated = update(codigos.filter(id.nullable().eq(unique_id)));
37 +             let updated = update(codigos);
   |


```

**Reply by weiznich** [reply, marked as answer]
Thanks for providing the update. The issue here is that your `codigos` table appears twice in your query, once in your subquery and once in your update. That means the name of that table would be ambiguous for the database system, therefore diesel disallows these kinds of constructs at compile time. You can easily fix that by using an alias for one of the table, for example as following:

```rust

            let codigos_alias = diesel::alias!(codigos as c);
            let unique_id = codigos_alias
                .filter(codigos_alias.field(codigos::expired).eq(false))
                .filter(codigos_alias.field(codigos::expires_at).lt(date(now)))
                .select(codigos_alias.field(codigos::id))
                .for_no_key_update()
                .skip_locked()
                .single_value();

            let updated = update(codigos::dsl::codigos)
                .filter(codigos::id.nullable().eq(unique_id))
                .set(codigos::expired.eq(true))
                .returning(codigos::id)
                .get_result(&mut conn)
                .await
                .optional()?;
```

**Reply by LtdJorge** [reply]
Thank you very much, Georg. I knew it had to be something simple that I was not seeing. Again, thank you.
